sophiread/SophireadCLI/src/sophiread_core.cpp

```diff
@@ -0,0 +1,382 @@
+ */
+std::vector<std::vector<std::vector<unsigned int>>> timedCreateTOFImages(
+    const std::vector<TPX3>& batches, 
+    double super_resolution, 
+    const std::vector<double>& tof_bin_edges,
+    const std::string& mode) {
+    
+    auto start = std::chrono::high_resolution_clock::now();
+
+    // Initialize the TOF images container
+    std::vector<std::vector<std::vector<unsigned int>>> tof_images(tof_bin_edges.size() - 1);
+
+    // Sanity checks
+    if (tof_bin_edges.size() < 2) {
+        spdlog::error("Invalid TOF bin edges: at least 2 edges are required");
+        return {};
+    }
+    if (batches.empty()) {
+        spdlog::error("No batches to process");
+        return tof_images;
+    }
+    
+    // Calculate the dimensions of each 2D histogram based on super_resolution
+    // one chip: 0-255 pixel pos
+    // gap: 5
+    // total: 0-255 + 5 + 0-255 -> 517 (TPX3@VENUS only)
+    int dim_x = static_cast<int>(517 * super_resolution);
+    int dim_y = static_cast<int>(517 * super_resolution);
+
+    spdlog::debug("Creating TOF images with dimensions: {} x {}", dim_x, dim_y);
+    spdlog::debug("tof_bin_edges size: {}", tof_bin_edges.size());
+    if (!tof_bin_edges.empty()) {
+        spdlog::debug("First bin edge: {}, Last bin edge: {}", tof_bin_edges.front(), tof_bin_edges.back());
+    }
+    
+    // Initialize each TOF bin's 2D histogram
+    for (auto& tof_image : tof_images) {
+        tof_image.resize(dim_y, std::vector<unsigned int>(dim_x, 0));
+    }
+
+    // Process neutrons from all batches
+    size_t total_entries = 0;
+    size_t binned_entries = 0;
+
+    for (size_t batch_index = 0; batch_index < batches.size(); ++batch_index) {
+        const auto& batch = batches[batch_index];
+        spdlog::debug("Processing batch {}", batch_index);
+
+        std::vector<const IPositionTOF*> entries;
+        if (mode == "hit") {
+            entries.reserve(batch.hits.size());
+            for (const auto& hit : batch.hits) {
+                entries.push_back(static_cast<const IPositionTOF*>(&hit));
+            }
+        } else {
+            entries.reserve(batch.neutrons.size());
+            for (const auto& neutron : batch.neutrons) {
+                entries.push_back(static_cast<const IPositionTOF*>(&neutron));
+            }
+        }
+        
+        if (entries.empty()) {
+            spdlog::debug("Batch {} is empty", batch_index);
+            continue;
+        }
+
+        for (const auto& entry : entries) {
+            total_entries++;
+            double tof_ns = entry->iGetTOF_ns();
+
+            // Find the correct TOF bin
+            // NOTE: tof_bin_edges are in sec, and tof_ns are in nano secs
+            spdlog::debug("tof_ns: {}, tof_ns/1e9: {}", tof_ns, tof_ns/1e9);
+            auto it = std::lower_bound(tof_bin_edges.begin(), tof_bin_edges.end(), tof_ns/1e9);
+            if (it != tof_bin_edges.begin()) {
```

Comment:
```suggestion
            if (const auto it std::lower_bound(tof_bin_edges.cbegin(), tof_bin_edges.cend(), tof_s); it != tof_bin_edges.cbegin()) {
```
Does this need to check against `cend()` as well?

Reply:
adjusted following suggestions.